File: src/runtime/env.zig
const std = @import("std");

/// Temporary Zig facade over the C runtime environment ABI.
///
/// Contract constraints: lookup still uses the caller-provided
/// `std.process.Environ`, so tests and explicit child environments do not fall
/// back to process-global `getenv`. All trimming, boolean, numeric, and desktop
/// token behavior is delegated to `env.c`. Returned slices borrow the supplied
/// environment storage and must not outlive or survive mutation of that storage.
/// Delete this facade when the remaining Zig callers move to C.
const CSpan = extern struct {
    data: ?[*]const u8,
    length: usize,
};

const status_valid: i32 = 1;

extern fn shaula_env_value_slice(value: ?[*:0]const u8, out: *CSpan) i32;
extern fn shaula_env_value_trimmed(value: ?[*:0]const u8, out: *CSpan) i32;
extern fn shaula_env_value_flag(value: ?[*:0]const u8, out_value: *i32) i32;
extern fn shaula_env_value_unsigned_or_default(
    value: ?[*:0]const u8,
    max_value: u64,
    default_value: u64,
) u64;
extern fn shaula_env_first_desktop_token(value: CSpan, out: *CSpan) i32;

comptime {
    if (@sizeOf(i32) != 4) @compileError("runtime env status ABI requires 32-bit i32");
    if (@sizeOf(u64) != 8) @compileError("runtime env unsigned ABI requires 64-bit u64");
    if (@sizeOf(CSpan) != @sizeOf(?[*]const u8) + @sizeOf(usize)) {
        @compileError("runtime env span ABI layout changed");
    }
}

fn borrowedSlice(span: CSpan) []const u8 {
    const data = span.data orelse unreachable;
    return data[0..span.length];
}

fn cValue(environ: std.process.Environ, key: []const u8) ?[*:0]const u8 {
    const value = environ.getPosix(key) orelse return null;
    return value.ptr;
}

pub fn slice(environ: std.process.Environ, key: []const u8) ?[]const u8 {
    var result: CSpan = .{ .data = null, .length = 0 };
    if (shaula_env_value_slice(cValue(environ, key), &result) != status_valid) {
        return null;
    }
    return borrowedSlice(result);
}

pub fn trimmed(environ: std.process.Environ, key: []const u8) ?[]const u8 {
    var result: CSpan = .{ .data = null, .length = 0 };
    if (shaula_env_value_trimmed(cValue(environ, key), &result) != status_valid) {
        return null;
    }
    return borrowedSlice(result);
}

pub fn flag(environ: std.process.Environ, key: []const u8) ?bool {
    var result: i32 = 0;
    if (shaula_env_value_flag(cValue(environ, key), &result) != status_valid) {
        return null;
    }
    return result != 0;
}

pub fn flagEnabled(environ: std.process.Environ, key: []const u8) bool {
    return flag(environ, key) orelse false;
}

pub fn unsignedOrDefault(comptime Int: type, environ: std.process.Environ, key: []const u8, default_value: Int) Int {
    const int_info = switch (@typeInfo(Int)) {
        .int => |info| info,
        else => @compileError("unsignedOrDefault requires an unsigned integer type"),
    };
    comptime {
        if (int_info.signedness != .unsigned or int_info.bits > 64) {
            @compileError("unsignedOrDefault supports unsigned integer types up to 64 bits");
        }
    }

    const parsed = shaula_env_value_unsigned_or_default(
        cValue(environ, key),
        @intCast(std.math.maxInt(Int)),
        @intCast(default_value),
    );
    return @intCast(parsed);
}

pub fn firstDesktopToken(value: []const u8) ?[]const u8 {
    var result: CSpan = .{ .data = null, .length = 0 };
    const input: CSpan = .{ .data = value.ptr, .length = value.len };
    if (shaula_env_first_desktop_token(input, &result) != status_valid) {
        return null;
    }
    return borrowedSlice(result);
}

test "facade preserves supplied environment lookup and borrowed lifetime" {
    var map = std.process.Environ.Map.init(std.testing.allocator);
    defer map.deinit();
    try map.put("FIRST", "first");
    try map.put("SECOND", "second");
    try map.put("EMPTY", "");
    try map.put("PADDED", " \tvalue\r\n");
    try map.put("FLAG", "YeS");
    try map.put("U8_MAX", "255");

    const block = try map.createPosixBlock(std.testing.allocator, .{});
    defer block.deinit(std.testing.allocator);
    const environ: std.process.Environ = .{ .block = block };

    const first = slice(environ, "FIRST").?;
    try std.testing.expectEqualStrings("second", slice(environ, "SECOND").?);
    try std.testing.expectEqualStrings("first", first);
    try std.testing.expectEqualStrings("", slice(environ, "EMPTY").?);
    try std.testing.expect(trimmed(environ, "EMPTY") == null);
    try std.testing.expectEqualStrings("value", trimmed(environ, "PADDED").?);
    try std.testing.expectEqual(true, flag(environ, "FLAG").?);
    try std.testing.expectEqual(@as(u8, 255), unsignedOrDefault(u8, environ, "U8_MAX", 7));
    try std.testing.expect(slice(environ, "MISSING") == null);
}

File: src/runtime/env.c
#include "env.h"

#include <string.h>

static void clear_span(ShaulaEnvSpan *span) {
  if (span == NULL) {
    return;
  }
  span->data = NULL;
  span->length = 0;
}

static int is_trim_byte(unsigned char value) {
  return value == (unsigned char)' ' || value == (unsigned char)'\t' ||
         value == (unsigned char)'\r' || value == (unsigned char)'\n';
}

static ShaulaEnvSpan trim_span(ShaulaEnvSpan value) {
  size_t start = 0;
  size_t end = value.length;

  while (start < end && is_trim_byte((unsigned char)value.data[start])) {
    start += 1;
  }
  while (end > start && is_trim_byte((unsigned char)value.data[end - 1])) {
    end -= 1;
  }

  return (ShaulaEnvSpan){value.data + start, end - start};
}

static unsigned char ascii_lower(unsigned char value) {
  if (value >= (unsigned char)'A' && value <= (unsigned char)'Z') {
    return (unsigned char)(value + ((unsigned char)'a' - (unsigned char)'A'));
  }
  return value;
}

static int span_equals_ascii_case_insensitive(ShaulaEnvSpan value,
                                              const char *expected) {
  size_t expected_length = strlen(expected);
  size_t index;

  if (value.length != expected_length) {
    return 0;
  }

  for (index = 0; index < value.length; index += 1) {
    if (ascii_lower((unsigned char)value.data[index]) !=
        ascii_lower((unsigned char)expected[index])) {
      return 0;
    }
  }

  return 1;
}

ShaulaEnvStatus shaula_env_value_slice(const char *value, ShaulaEnvSpan *out) {
  if (out == NULL) {
    return SHAULA_ENV_STATUS_INVALID;
  }

  clear_span(out);
  if (value == NULL) {
    return SHAULA_ENV_STATUS_MISSING;
  }

  out->data = value;
  out->length = strlen(value);
  return SHAULA_ENV_STATUS_VALID;
}

ShaulaEnvStatus shaula_env_value_trimmed(const char *value,
                                         ShaulaEnvSpan *out) {
  ShaulaEnvSpan raw;
  ShaulaEnvSpan trimmed;
  ShaulaEnvStatus status;

  if (out == NULL) {
    return SHAULA_ENV_STATUS_INVALID;
  }

  clear_span(out);
  status = shaula_env_value_slice(value, &raw);
  if (status != SHAULA_ENV_STATUS_VALID) {
    return status;
  }

  trimmed = trim_span(raw);
  if (trimmed.length == 0) {
    return SHAULA_ENV_STATUS_MISSING;
  }

  *out = trimmed;
  return SHAULA_ENV_STATUS_VALID;
}

ShaulaEnvStatus shaula_env_value_flag(const char *value, int32_t *out_value) {
  ShaulaEnvSpan trimmed;
  ShaulaEnvStatus status;

  if (out_value == NULL) {
    return SHAULA_ENV_STATUS_INVALID;
  }

  *out_value = 0;
  status = shaula_env_value_trimmed(value, &trimmed);
  if (status != SHAULA_ENV_STATUS_VALID) {
    return status;
  }

  if ((trimmed.length == 1 && trimmed.data[0] == '1') ||
      span_equals_ascii_case_insensitive(trimmed, "true") ||
      span_equals_ascii_case_insensitive(trimmed, "yes")) {
    *out_value = 1;
    return SHAULA_ENV_STATUS_VALID;
  }

  if ((trimmed.length == 1 && trimmed.data[0] == '0') ||
      span_equals_ascii_case_insensitive(trimmed, "false") ||
      span_equals_ascii_case_insensitive(trimmed, "no")) {
    return SHAULA_ENV_STATUS_VALID;
  }

  return SHAULA_ENV_STATUS_INVALID;
}

uint64_t shaula_env_value_unsigned_or_default(const char *value,
                                              uint64_t max_value,
                                              uint64_t default_value) {
  ShaulaEnvSpan trimmed;
  ShaulaEnvStatus status;
  size_t index = 0;
  int negative = 0;
  uint64_t parsed = 0;

  status = shaula_env_value_trimmed(value, &trimmed);
  if (status != SHAULA_ENV_STATUS_VALID) {
    return default_value;
  }

  if (trimmed.data[index] == '+' || trimmed.data[index] == '-') {
    negative = trimmed.data[index] == '-';
    index += 1;
    if (index == trimmed.length) {
      return default_value;
    }
  }

  if (trimmed.data[index] == '_' || trimmed.data[trimmed.length - 1] == '_') {
    return default_value;
  }

  for (; index < trimmed.length; index += 1) {
    unsigned char byte = (unsigned char)trimmed.data[index];
    uint64_t digit;

    if (byte == (unsigned char)'_') {
      continue;
    }
    if (byte < (unsigned char)'0' || byte > (unsigned char)'9') {
      return default_value;
    }

    digit = (uint64_t)(byte - (unsigned char)'0');
    if (digit > max_value || parsed > (max_value - digit) / UINT64_C(10)) {
      return default_value;
    }
    parsed = parsed * UINT64_C(10) + digit;
  }

  if (negative && parsed != 0) {
    return default_value;
  }

  return parsed;
}

ShaulaEnvStatus shaula_env_first_desktop_token(ShaulaEnvSpan value,
                                               ShaulaEnvSpan *out) {
  size_t token_start = 0;
  size_t index;

  if (out == NULL) {
    return SHAULA_ENV_STATUS_INVALID;
  }

  clear_span(out);
  if (value.data == NULL) {
    return SHAULA_ENV_STATUS_MISSING;
  }

  for (index = 0; index <= value.length; index += 1) {
    int at_end = index == value.length;
    int at_separator =
        !at_end && (value.data[index] == ':' || value.data[index] == ';');

    if (at_end || at_separator) {
      ShaulaEnvSpan token = trim_span(
          (ShaulaEnvSpan){value.data + token_start, index - token_start});
      if (token.length > 0) {
        *out = token;
        return SHAULA_ENV_STATUS_VALID;
      }
      token_start = index + 1;
    }
  }

  return SHAULA_ENV_STATUS_MISSING;
}
